### DISCUSSION TITLE
Waku errors with jotai-ssr

### DISCUSSION THREAD

**Discussion by dai-shi** [wakujs/waku#863] [General]
@akkadaska reported several issues in https://github.com/pmndrs/jotai/discussions/2692

> Additionally, a Waku-specific error not encountered with Next.js and Remix occurs:
> 
> * **Hydration Mismatch Error:**  Directly accessing pages that utilize `useHydrateAtoms` or `HydrationBoundary` without `RenderingBoundary` leads to a hydration mismatch error: "Hydration failed because the server rendered HTML didn't match the client. As a result, this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used:..."
> 
> Finally, an issue unrelated to Jotai occurs when importing modules containing the `'use client';` directive (like `Provider`, `HydrationBoundary`) from server components (`pages/**/*.tsx`):
> 
> * **Server-Side Import Error:**  Importing these modules from server components causes a server error. For instance, importing `Provider` in `pages/index.tsx` results in this error:
> 
> ```
> Cannot find module '/node_modules/jotai/esm/react.mjs' imported from /path-to-project/node_modules/waku/dist/lib/middleware/dev-server-impl.js
> ```

**Comment by dai-shi** [comment]
Re: https://github.com/pmndrs/jotai/discussions/2692#discussioncomment-10512268

https://stackblitz.com/edit/github-bbtjui?file=src%2Fcomponents%2Fprovider.tsx,src%2Fpages%2F_layout.tsx doesn't cause any errors.
@akkadaska Can you help making a small reproduction?

**Reply by akkadaska** [reply]
# Server-Side Import Error

@dai-shi 
Here is a small reproduction of **Server-Side Import Error**: importing `RenderingBoundary` from `jotai-ssr` in `_layout.tsx`
https://stackblitz.com/edit/github-bbtjui-d628pt?file=src%2Fpages%2F_layout.tsx

This error occurs with more famous library e.g. Tanstack Query.
https://stackblitz.com/edit/github-bbtjui-byy1se?file=src%2Fpages%2F_layout.tsx

HydrationBoundary from @tanstack/react-query has also 'use client' directive.
https://github.com/TanStack/query/blob/f8ad97355c502f97eace180e418d0fa27986f603/packages/react-query/src/HydrationBoundary.tsx#L3

Note that at least importing **library** that has `'use client'` directive causes an error, and creating a component that has `'use client'` in the project is works well.

**Reply by akkadaska** [reply]
React document says this should work.

> If these libraries have been updated to be compatible with React Server Components, then they will already include `'use client'` markers of their own, allowing you to use them directly from your Server Components.

https://react.dev/reference/rsc/use-client#using-third-party-libraries

**Reply by akkadaska** [reply]
@dai-shi 
And **Hydration Mismatch Error** was caused by my mistake of implementation. With correct one, this works well.
https://github.com/akkadaska/jotai-waku-link-navigation/commit/4a517378ddfdf97bf58d6aefbcc9943c150b46ad

I am very sorry for my mistake. Waku seems to have only "Server-Side Import Error" I think.

**Reply by dai-shi** [reply]
> React document says this should work.

Yeah, I know it, and Jotai does it too.

**Reply by dai-shi** [reply]
https://stackblitz.com/edit/github-bbtjui-q5yjrk?file=src%2Fpages%2F_layout.tsx

Reproduced with Jotai too. Oh, I wasn't aware of this. Let's fix.

**Reply by dai-shi** [reply]
https://stackblitz.com/edit/github-y5y7tq-fsxgko?file=src%2Fentries.tsx

Can reproduce without waku/router.

**Reply by dai-shi** [reply]
Hmm, e2e test can't reproduce it.

**Reply by dai-shi** [reply]
PR 👉 #864

**Reply by dai-shi** [reply]
#864 fixes it: https://stackblitz.com/edit/github-y5y7tq-bxd7mm?file=package.json

**Comment by dai-shi** [comment]
https://github.com/dai-shi/waku/discussions/863#discussioncomment-10525494

> @dai-shi And **Hydration Mismatch Error** was caused by my mistake of implementation. With correct one, this works well. [akkadaska/jotai-waku-link-navigation@4a51737](https://github.com/akkadaska/jotai-waku-link-navigation/commit/4a517378ddfdf97bf58d6aefbcc9943c150b46ad)
> 
> I am very sorry for my mistake. Waku seems to have only "Server-Side Import Error" I think.

Thanks for confirming. That's a relief.
